File: src/main/java/com/devxpress/Tree.java
package com.devxpress;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;
import java.util.stream.Collectors;

public class Tree<T> {

    private final Node<T> root;

    public Tree(T data) {
        this.root = new Node<>(data);
    }

    public Node<T> getRoot() {
        return this.root;
    }

    public void traverseBF(Consumer<Node<T>> func) {
        List<Node<T>> nodeList = new ArrayList<>();
        nodeList.add(this.root);

        while (!nodeList.isEmpty()) {
            Node<T> node = nodeList.remove(0);
            // Add to end of list
            nodeList.addAll(node.children);
            func.accept(node);
        }
    }

    public void traverseDF(Consumer<Node<T>> func) {
        List<Node<T>> nodeList = new ArrayList<>();
        nodeList.add(this.root);

        while (!nodeList.isEmpty()) {
            Node<T> node = nodeList.remove(0);
            // Add to start of list
            nodeList.addAll(0, node.children);
            func.accept(node);
        }
    }

    static class Node<T> {
        private final T data;
        private final List<Node<T>> children;

        Node(T data) {
            this.data = data;
            this.children = new ArrayList<>();
        }

        T getData() {
            return this.data;
        }

        List<Node<T>> getChildren() {
            return new ArrayList<>(this.children);
        }

        void add(T data) {
            this.children.add(new Node<>(data));
        }

        void remove(T data) {
            this.children.removeAll(this.children.stream().
                                filter(e -> e.data.equals(data)).
                                collect(Collectors.toList()));
        }

        static int[] levelWidth(Node root) {
            List<Integer> counts = new ArrayList<>();

            List<Node> currentLevel = new ArrayList<>();
            List<Node> nextLevel = new ArrayList<>();

            // Prime the current level for the root node
            currentLevel.add(root);
            counts.add(0);

            while (!currentLevel.isEmpty() || !nextLevel.isEmpty()) {
                // Remove first node at current level
                Node node = currentLevel.remove(0);
                // Increment the last counter
                counts.set(counts.size() - 1, counts.get(counts.size() - 1) + 1);
                // Stick children in next level list
                nextLevel.addAll(node.children);

                if (currentLevel.isEmpty()) {
                    // Check whether there is a next level
                    if (!nextLevel.isEmpty()) {
                        // Start of the next level so add counter
                        counts.add(0);
                        // Now swap current and next
                        List<Node> temp = currentLevel;
                        currentLevel = nextLevel;
                        nextLevel = temp;
                    }
                }
            }

            return counts.stream().mapToInt(Integer::intValue).toArray();
        }
    }
}

File: src/main/resources/js/tree.js
function Tree() {
    this.root = null;
}

Tree.prototype.traverseBF = function(f) {
    if (!this.root) {
        throw "Traversal of empty Tree is not possible";
    }

    var nodeArr = [ this.root ];

    while (nodeArr.length > 0) {
        // Remove first element from array
        var node = nodeArr.shift();

        // Add each of node's children and add to end of array
        node.children.forEach(function(e) {
            nodeArr.push(e);
        });

        // Apply the function passed
        f(node);
    }
};

Tree.prototype.traverseDF = function(f) {
    if (!this.root) {
        throw "Traversal of empty Tree is not possible";
    }

    var nodeArr = [ this.root ];

    while (nodeArr.length > 0) {
        // Remove first element from array
        var node = nodeArr.shift();

        // Add each of node's children and add to start of array
        // Have to iterate backwards to maintain order
        for (var i = node.children.length - 1; i >= 0; i--) {
            nodeArr.unshift(node.children[i]);
        }

        // Apply the function passed
        f(node);
    }
};

function Node(data) {
    this.data = data;
    this.children = [];
}

Node.prototype.add = function(data) {
    this.children.push(new Node(data));
};

Node.prototype.remove = function(data) {
    this.children = this.children.filter(function(e) {
        return e.data !== data;
    });
};

Node.prototype.levelWidth = function(root) {
    var counts = [], currentLevel = [], nextLevel = [];

    // Prime the current level for the root node
    currentLevel.push(root);
    counts.push(0);

    while (currentLevel.length || nextLevel.length) {
        // Remove first node at current level
        var node = currentLevel.shift();
        // Increment the last counter
        counts[counts.length - 1]++;
        // Stick children in next level array
        node.children.forEach(function(node) { nextLevel.push(node) });

        if (!currentLevel.length) {
            // Check whether there is a next level
            if (nextLevel.length) {
                // Start of the next level, so add a new counter
                counts.push(0);
                // Now swap current and next arrays
                var temp = currentLevel;
                currentLevel = nextLevel;
                nextLevel = temp;
            }
        }
    }

    return counts;
}

function testTree() {
    var results = [];
    var t = new Tree();
    t.root = new Node('a');
    t.root.add('b');
    t.root.add('c');
    t.root.add('d');

    t.root.children[0].add('e');
    t.root.children[0].add('f');
    t.root.children[0].add('g');
    t.root.children[2].add('h');
    t.root.children[2].add('i');

    print(JSON.stringify(t));

    t.traverseBF(function(node) {
        results.push(node.data);
    });

    print("Traverse BF Results: " + results.toString());

    results = [];

    t.traverseDF(function(node) {
        results.push(node.data);
    });

    print("Traverse DF Results: " + results.toString());

    print("Level Width: " + Node.prototype.levelWidth(t.root));
}